README.md
                                                                   
       _________                                           .___      
       \_   ___ \  ____   _____   _____ _____    ____    __| _/____  
       /    \  \/ /  _ \ /     \ /     \\__  \  /    \  / __ |/  _ \ 
       \     \___(  <_> )  Y Y  \  Y Y  \/ __ \|   |  \/ /_/ (  <_> )
        \______  /\____/|__|_|  /__|_|  (____  /___|  /\____ |\____/ 
               \/             \/      \/     \/     \/      \/       
                        C O M P L E T E  M A N D I A N T                    
                             O F F E N S I V E   V M                        
                                  Version 2021.2                                 
                              commandovm@fireeye.com
               _____________________________________________________          

                                    Created by                                
                             Jake Barteaux @day1player                              
                                 Mandiant Red Team  
                          Blaine Stancill @MalwareMechanic                           
                                    Nhan Huynh   
                     FireEye Labs Advanced Reverse Engineering                            
______________________________________________________________________________ 

<p align="center">
  <img width="300" src="https://github.com/fireeye/commando-vm/blob/master/Commando.png?raw=true" alt="Commando VM"/>
</p>  

Welcome to CommandoVM - a fully customizable, Windows-based security distribution for penetration testing and red teaming.

For detailed install instructions or more information please see our [blog](https://www.fireeye.com/blog/threat-research/2019/08/commando-vm-customization-containers-kali.html)

Installation (Install Script)
=============================

Requirements
------------
* Windows 10 1803, 1809, 1903, 1909, 2004, 20H2, or 21H1
> Insider Preview editions of Windows are not supported
* 60 GB Hard Drive
* 2 GB RAM

Recommended
-----------
* Windows 10 21H1
* 80+ GB Hard Drive
* 4+ GB RAM
* 2 network adapters

Instructions
============
Pre-Install Procedures
----------------------
You MUST disable Windows Defender for a smooth install. The best way to accomplish this is through Group Policy
In Windows versions 1909 and higher, Tamper Protection was added.
**Tamper Protection must be disabled, otherwise Group Policy settings are ignored.**

1. Open Windows Security (type `Windows Security` in the search box)
2. Virus & threat protection > Virus & threat protection settings > Manage settings
3. Switch `Tamper Protection` to `Off` 
> It is not necessary to change any other setting (`Real Time Protection`, etc.)

> **Important.** Tamper Protection must be disabled before changing Group Policy settings.

To permanently disable Real Time Protection

1. Open Local Group Policy Editor (type `gpedit` in the search box)
2. Computer Configuration > Administrative Templates > Windows Components > Microsoft Defender Antivirus > Real-time Protection
3. Enable `Turn off real-time protection`
4. **Reboot**
> Make sure to **reboot** before making the next change

To permanently disable Microsoft Defender:

1. Open Local Group Policy Editor (type `gpedit` in the search box)
2. Computer Configuration > Administrative Templates > Windows Components > Microsoft Defender Antivirus
3. Enable `Turn off Microsoft Defender Antivirus`
4. **Reboot**

  [1]: https://stackoverflow.com/questions/62174426/how-to-permanently-disable-windows-defender-real-time-protection-with-gpo

Standard install
----------------
1. Create and configure a new Windows Virtual Machine
  > Ensure VM is updated completely. You may have to check for updates, reboot, and check again until no more remain 
2.  Take a snapshot of your machine!
3.  Download and copy `install.ps1` on your newly configured machine. 
4.  Open PowerShell as an Administrator
5.  Unblock the install file by running `Unblock-File .\install.ps1`
6.  Enable script execution by running `Set-ExecutionPolicy Unrestricted -f`
7.  Finally, execute the installer script as follows:
   * `.\install.ps1`
   * You can also pass your password as an argument: `.\install.ps1 -password <password>`
  
The script will set up the Boxstarter environment and proceed to download and install the Commando VM environment. You will be prompted for the administrator password in order to automate host restarts during installation. If you do not have a password set, hitting enter when prompted will also work.

Custom install
--------------
> Please see our [custom profiles](https://github.com/fireeye/commando-vm/tree/master/Profiles) for more custom install options or create your own following the instructions below.
1.	Download the zip from https://github.com/fireeye/commando-vm into your Downloads folder.
2.	Decompress the zip and edit the `${Env:UserProfile}\Downloads\commando-vm-master\commando-vm-master\profile.json` file by removing tools or adding tools in the “packages” section. You can add any package listed in our [package list](https://github.com/fireeye/commando-vm/blob/master/packages.csv) or any package from the [chocolatey repository](https://chocolatey.org/packages).
3.	Open an administrative PowerShell window and enable script execution.
`Set-ExecutionPolicy Unrestricted -f`
4.	Change to the unzipped project directory.
`cd ${Env:UserProfile}\Downloads\commando-vm-master\commando-vm-master\`
5.  Unblock the install file by running `Unblock-File .\install.ps1`
6.	Take a snapshot of your machine!
7. Execute the install with the `-profile_file` argument.
`.\install.ps1 -profile_file .\profile.json`

For more detailed instructions about custom installations, see our [blog](https://www.fireeye.com/blog/threat-research/2019/08/commando-vm-customization-containers-kali.html)

Installing a new package
========================

Commando VM uses the [Chocolatey](https://chocolatey.org/) Windows package manager. It is easy to install a new package. For example, enter the following command as Administrator to deploy Github Desktop on your system:

    cinst github

You can find packages to install from our [package list](https://github.com/fireeye/commando-vm/blob/master/packages.csv), which hosts more than just pentesting tools, or from the [chocolatey repository](https://chocolatey.org/packages).


Staying up to date
==================

Type the following command to update all of the packages to the most recent version:

    cup all


Available Tools
===============

### Active Directory Tools
- Remote Server Administration Tools (RSAT)
- SQL Server Command Line Utilities
- Sysinternals

### Command & Control
- Covenant
- WMImplant
- WMIOps

### Developer Tools
- Dep
- Git
- Go
- Java
- Python 2
- Python 3 (default)
- Ruby
- Ruby Devkit
- Visual Studio 2017 Build Tools (Windows 10)
- Visual Studio Code

### Docker
- Amass
- SpiderFoot

### Evasion
- CheckPlease
- Demiguise
- DefenderCheck
- DotNetToJScript
- Invoke-CradleCrafter
- Invoke-DOSfuscation
- Invoke-Obfuscation
- Invoke-Phant0m
- Not PowerShell (nps)
- PS>Attack
- PSAmsi
- Pafishmacro
- PowerLessShell
- PowerShdll
- StarFighters
- SysWhispers

### Exploitation
- ADAPE-Script
- API Monitor
- CrackMapExec
- CrackMapExecWin
- DAMP
- Dumpert
- EvilClippy
- Exchange-AD-Privesc
- FuzzySec's PowerShell-Suite
- FuzzySec's Sharp-Suite
- GadgetToJScript
- Generate-Macro
- GhostPack
  - Rubeus
  - SafetyKatz
  - Seatbelt
  - SharpDPAPI
  - SharpDump
  - SharpRoast
  - SharpUp
  - SharpWMI
- GoFetch
- Impacket
- Invoke-ACLPwn
- Invoke-DCOM
- Invoke-PSImage
- Invoke-PowerThIEf
- Juicy Potato
- Kali Binaries for Windows
- LuckyStrike
- MetaTwin
- Metasploit
- Mr. Unikod3r's RedTeamPowershellScripts
- NetshHelperBeacon
- Nishang
- Orca
- PSBits
- PSReflect
- PowerLurk
- PowerPriv
- PowerSploit
- PowerUpSQL
- PrivExchange
- RottenPotatoNG
- Ruler
- SharpClipHistory
- SharpExchangePriv
- SharpExec
- SpoolSample
- SharpSploit
- ThreadContinue
- TikiTorch
- UACME
- impacket-examples-windows
- vssown
- Vulcan

### Information Gathering
- ADACLScanner
- ADExplorer
- ADOffline
- ADRecon
- BeRoot
- BloodHound
- BloodHound-Custom-Queries (Hausec)
- dnsrecon
- FOCA
- Get-ReconInfo
- GoBuster
- GoWitness
- Net-GPPPassword
- NetRipper
- Nmap

changelog.md
## 2021.2 - July 14 2021
- Require users to disable Defender before installing
  * Too many issues arise from Defender magically turning itself back on. Disabling defender with the preconfig script has been unreliable since Win10 1909.
- removed update requirement (legacy  requirement for Windows 7. Windows 7 support was  removed last year)
- Added support for Windows 10 20H2 and 21H1
- Removed vcpython27 #204
- updated proxycap install args #203, #200. #196
- updated sqlitebrowser.fireeye to remove newly created desktop shortcuts #200
- Closed issues #203, #204, #202, #200, #196, #195, #192, #191, #190, #189, #188, #186, #185, #184, #177, #175, #174, #170, #169, #160, #134, #133

## 2020.2 - June 17 2020
- Added support for Windows 10 2004
- Corrected syspin verb #124
- Removed WSL from default install #146, #120
- Removed Hyper V from default install #146, #120
- Removed Kali from default install #95, #120
- Removed Docker from default install #95, #120
- Created wsl.fireeye package #95, #120
- Created hyperv.fireeye package #95, #120 
- Created multiple install profiles #95, #120
  - Default, NestedV, Lite, Full, and Developer
- Removed some dependencies causing %PATH% variable to be truncated to 1024 characters #141
- Added logic to help speed up install
- Removed dependency on custom libraries file #131
- Added custom logging for installation of packages #70
- Fixed bug in autohotkey script for unhardening of UNC paths #68
- Updated Readme #140

## 2020.1 - March 3 2020
- added logic to attempt automatically updating system #88
- added qbittorrent #88
- added dbeaver #88
- added hfsexplorer #88
- added lockhunter #88
- fixed typo for PwnedPasswordsNTLM in packages.json #101
- added BeRoot
- added BloodHound Custom Queries - Hausec
- added Dumpert
- added Recon-AD
- added Net-GPPPassword
- added Gadget2JScript
- added OffensiveCSharp - matterpreter
- added powercat
- added Privesc - enjoiz
- added PSBits
- added ThreadContinue
- added SysWhispers
- added TikiTorch
- added Virus Total Uploader #88
- added NirLauncher #88
- added SimpleDnsCrypt #88
- added Tor Browser #88
- added HeidiSQL #88
- added HTTP File Server #88
- Removed support for Windows 7 (install should still work, but is no longer maintained)
- improved error handling for commandovm.win10.config.fireeye package
- updated commandovm.win10.preconfig.fireeye
- Fixed taskbar pinning on 1903 (still not working for 1909)
- Updated install instructions on readme
- Updated ~45 packages
- Removed Watson binaries (static binaries are not updated in this repo)
- Removed PoshC2 (deprecated, will look at supporting PoshC2 Python)
- Removed Covenant (will support again in a future release)
- Removed Elite (deprecated)

## 2.0 - August 5 2019
- Added Kali Linux https://www.kali.org
- Added Docker https://www.docker.com #88
- Added SpiderFoot https://github.com/smicallef/spiderfoot #84
- Added Amass https://github.com/OWASP/Amass
- Added customization support #42, #25 

## 1.3 - June 28 2019
- Added RottenPotatoNG https://github.com/breenmachine/RottenPotatoNG #63
- Added Juicy Potato https://github.com/ohpe/juicy-potato #63, #64
- Added Watson https://github.com/rasta-mouse/Watson #64
- Added PwndPasswordsNTLM https://github.com/JacksonVD/PwnedPasswordsNTLM #67
- Added FOCA https://github.com/JacksonVD/PwnedPasswordsNTLM #71 
- Added Vulcan https://github.com/praetorian-code/vulcan
- Added SharpClipHistory https://github.com/mwrlabs/SharpClipHistory
- Added NetRipper https://github.com/NytroRST/NetRipper
- Added RobotsDisallowed https://github.com/danielmiessler/RobotsDisallowed
- Added Probable-Wordlists https://github.com/berzerk0/Probable-Wordlists
- Added SharpSploit https://github.com/cobbr/SharpSploit
- Changed WinRM configuration #65
- Un-hardened UNC file paths #68
- Fixed install issues with Covenant #61, #76

## 1.2 - May 31 2019
- Added recommended hardware settings #20, #17
- Added DomainPasswordSpray https://github.com/dafthack/DomainPasswordSpray #2
- Added GoBuster https://github.com/OJ/gobuster #39
- Added Wfuzz https://github.com/xmendez/wfuzz #40
- Added Notepad++ #30
- Added TextFX plugin for Notepad++

install.ps1
###########################################
#
# CommandoVM Installation Script
#
# To execute this script:
# 1) Open powershell window as administrator
# 2) Allow script execution by running command "Set-ExecutionPolicy Unrestricted"
# 3) Unblock the install script by running "Unblock-File .\install.ps1"
# 4) Execute the script by running ".\install.ps1"
#
###########################################
param (
  [string]$password = "",  
  [string]$profile_file = $null, 
  [bool]$nochecks = $false
)


function Set-EnvironmentVariableWrap([string] $key, [string] $value)
{
<#
.SYNOPSIS
	Set the environment variable for all process, user and system wide scopes
.OUTPUTS
	True on success | False on error
#>
  try {
	  [Environment]::SetEnvironmentVariable($key, $value)
	  [Environment]::SetEnvironmentVariable($key, $value, 1)
	  [Environment]::SetEnvironmentVariable($key, $value, 2)
	
	  $rc = $true
	} catch {
		$rc = $false
	}
	$rc
}


function ConvertFrom-Json([object] $item) {
<#
.SYNOPSIS
  Convert a JSON string into a hash table

.DESCRIPTION
  Convert a JSON string into a hash table, without any validation

.OUTPUTS
  [hashtable] or $null
#>
  Add-Type -Assembly system.web.extensions
  $ps_js = New-Object system.web.script.serialization.javascriptSerializer

  try {
    $result = $ps_js.DeserializeObject($item)
  } catch {
    $result = $null
  }
  
  # Cast dictionary to hashtable
  [hashtable] $result
}


function ConvertTo-Json([object] $data) {
<#
.SYNOPSIS
  Convert a hashtable to a JSON string

.DESCRIPTION
  Convert a hashtable to a JSON string, without any validation

.OUTPUTS
  [string] or $null
#>
  Add-Type -Assembly system.web.extensions
  $ps_js = New-Object system.web.script.serialization.javascriptSerializer

  #The comma operator is the array construction operator in PowerShell
  try {
    $result = $ps_js.Serialize($data)
  } catch {
    $result = $null
  }
  
  $result
}


function Import-JsonFile {
<#
.DESCRIPTION
  Load a hashtable from a JSON file
  
.OUTPUTS
  [hashtable] or $null
#>
  param([string] $path)
  try {
    $json = Get-Content $path
    $result = ConvertFrom-Json $json
  } catch {
    $result = $null
  }
  
  $result
}


function Make-InstallerPackage($PackageName, $TemplateDir, $packages) {
	<#
	.SYNOPSIS
	Make a new installer package

	.DESCRIPTION
	Make a new installer package named installer. This package uses the custom packages.json file specified by the user.
	User can then call "Install-BoxStarterPackage installer" using the local repo.
	#>

	$PackageDir = Join-Path $BoxStarter.LocalRepo $PackageName
	if (Test-Path $PackageDir) {
		Remove-Item -Recurse -Force $PackageDir
	}

	$Tmp = [System.IO.Path]::GetTempFileName()
	Write-Host -ForegroundColor Green "packages file is" + $tmp
	ConvertTo-Json @{"packages" = $packages} | Out-File -FilePath $Tmp
	
  if ([System.IO.Path]::IsPathRooted($TemplateDir)) {
    $ToolsDir = Join-Path $TemplateDir "tools"
  } else {
	  $Here = Get-Location
	  $ToolsDir = Join-Path (Join-Path $Here $TemplateDir) "tools"
  }
  $Dest = Join-Path $ToolsDir "packages.json"

	Move-Item -Force -Path $Tmp -Destination $Dest
	New-BoxstarterPackage -Name $PackageName -Description "My Own Instalelr" -Path $ToolsDir
}

function installBoxStarter()
{
  <#
  .SYNOPSIS
  Install BoxStarter on the current system  
  .DESCRIPTION
  Install BoxStarter on the current system. Returns $true or $false to indicate success or failure. On
  fresh windows 7 systems, some root certificates are not installed and updated properly. Therefore,
  this funciton also temporarily trust all certificates before installing BoxStarter.  
  #> 
  

  # Try to install BoxStarter as is first, then fall back to be over trusing only if this step fails.
  try {
		iex ((New-Object System.Net.WebClient).DownloadString('https://boxstarter.org/bootstrapper.ps1')); get-boxstarter -Force
		return $true
	} catch {
	}
   
  # https://stackoverflow.com/questions/11696944/powershell-v3-invoke-webrequest-https-error
  # Allows current PowerShell session to trust all certificates
  # Also a good find: https://www.briantist.com/errors/could-not-establish-trust-relationship-for-the-ssltls-secure-channel/
  
  try {
  Add-Type @"
  using System.Net;
  using System.Security.Cryptography.X509Certificates;
  public class TrustAllCertsPolicy : ICertificatePolicy {
  	public bool CheckValidationResult(
  		ServicePoint srvPoint, X509Certificate certificate,
  		WebRequest request, int certificateProblem) {
  		return true;
  	}
  }
"@
  } catch {
    Write-Debug "Failed to add new type"
  }  
  try {
  	$AllProtocols = [System.Net.SecurityProtocolType]'Ssl3,Tls,Tls11,Tls12'
  } catch {
  	Write-Debug "Failed to find SSL type...1"
  }  
  try {
  	$AllProtocols = [System.Net.SecurityProtocolType]'Ssl3,Tls'
  } catch {
  	Write-Debug "Failed to find SSL type...2"
  }  
  $prevSecProtocol = [System.Net.ServicePointManager]::SecurityProtocol
  $prevCertPolicy = [System.Net.ServicePointManager]::CertificatePolicy  
  Write-Host "[+] Installing Boxstarter"
  # Become overly trusting
  [System.Net.ServicePointManager]::SecurityProtocol = $AllProtocols
  [System.Net.ServicePointManager]::CertificatePolicy = New-Object TrustAllCertsPolicy  
  # download and instal boxstarter
  iex ((New-Object System.Net.WebClient).DownloadString('https://boxstarter.org/bootstrapper.ps1')); get-boxstarter -Force  
  # Restore previous trust settings for this PowerShell session
  # Note: SSL certs trusted from installing BoxStarter above will be trusted for the remaining PS session
  [System.Net.ServicePointManager]::SecurityProtocol = $prevSecProtocol
  [System.Net.ServicePointManager]::CertificatePolicy = $prevCertPolicy
  return $true
}
function Wait-ForInstall ($seconds) {
  $doneDT = (Get-Date).AddSeconds($seconds)
  while($doneDT -gt (Get-Date)) {
      $secondsLeft = $doneDT.Subtract((Get-Date)).TotalSeconds
      $percent = ($seconds - $secondsLeft) / $seconds * 100
      Write-Progress -Activity "Please read install notes on console below" -Status "Beginning install in..." -SecondsRemaining $secondsLeft -PercentComplete $percent
      [System.Threading.Thread]::Sleep(500)
  }
  Write-Progress -Activity "Waiting" -Status "Beginning install..." -SecondsRemaining 0 -Completed
}

Write-Host "`n"
Write-Host " ____________________________________________________________________________ " -ForegroundColor Red 
Write-Host "|                                                                            |" -ForegroundColor Red 
Write-Host "|    "  -ForegroundColor Red -NoNewline; Write-Host "                  " -ForegroundColor Green -NoNewline; Write-Host "                                                      |" -ForegroundColor Red 
Write-Host "|        "  -ForegroundColor Red -NoNewline; Write-Host "_________                                           .___      " -ForegroundColor Green -NoNewline; Write-Host "      |" -ForegroundColor Red 
Write-Host "|        "  -ForegroundColor Red -NoNewline; Write-Host "\_   ___ \  ____   _____   _____ _____    ____    __| _/____  " -ForegroundColor Green -NoNewline; Write-Host "      |" -ForegroundColor Red 
Write-Host "|        "  -ForegroundColor Red -NoNewline; Write-Host "/    \  \/ /  _ \ /     \ /     \\__  \  /    \  / __ |/  _ \ " -ForegroundColor Green -NoNewline; Write-Host "      |" -ForegroundColor Red 
Write-Host "|        "  -ForegroundColor Red -NoNewline; Write-Host "\     \___(  <_> )  Y Y  \  Y Y  \/ __ \|   |  \/ /_/ (  <_> )" -ForegroundColor Green -NoNewline; Write-Host "      |" -ForegroundColor Red 
Write-Host "|        "  -ForegroundColor Red -NoNewline; Write-Host " \______  /\____/|__|_|  /__|_|  (____  /___|  /\____ |\____/ " -ForegroundColor Green -NoNewline; Write-Host "      |" -ForegroundColor Red 
Write-Host "|        "  -ForegroundColor Red -NoNewline; Write-Host "        \/             \/      \/     \/     \/      \/       " -ForegroundColor Green -NoNewline; Write-Host "      |" -ForegroundColor Red 
Write-Host "|                       C O M P L E T E  M A N D I A N T                     |" -ForegroundColor Red 
Write-Host "|                            O F F E N S I V E   V M                         |" -ForegroundColor Red 
Write-Host "|                                                                            |" -ForegroundColor Red 
Write-Host "|                                 Version 2021.2                             |" -ForegroundColor Red 
Write-Host "|                             commandovm@fireeye.com                         |" -ForegroundColor Red 
Write-Host "|____________________________________________________________________________|" -ForegroundColor Red 
Write-Host "|                                                                            |" -ForegroundColor Red 
Write-Host "|                                   Created by                               |" -ForegroundColor Red 
Write-Host "|                            Jake Barteaux @day1player                       |" -ForegroundColor Red 
Write-Host "|                               Proactive Services                           |" -ForegroundColor Red 
Write-Host "|                         Blaine Stancill @MalwareMechanic                   |" -ForegroundColor Red 
Write-Host "|                                   Nhan Huynh                               |" -ForegroundColor Red 
Write-Host "|                    FireEye Labs Advanced Reverse Engineering               |" -ForegroundColor Red 
Write-Host "|____________________________________________________________________________|" -ForegroundColor Red 
Write-Host ""
  
if ([string]::IsNullOrEmpty($profile_file)) {
  Write-Host "[+] No custom profile is provided..."
  $profile = $null
} else {
  Write-Host "[+] Using the following profile $profile_file"
  $profile = Import-JsonFile $profile_file
  if ($profile -eq $null) {
    Write-Error "Invaild configuration! Exiting..."
    exit 1
  }
  # Confirmation message
  Write-Warning "[+] You are using a custom profile and list of packages,"
  Write-Warning "[+] You will NOT receive new tools automatically when running choco update."

  if ($nochecks -eq $false) {
    Write-Host "[-] Do you want to continue? Y/N " -ForegroundColor Yellow -NoNewline
    $response = Read-Host
    if ($response -ne "Y") {
      Write-Host "[*] Exiting..." -ForegroundColor Red
      exit
    }
  }
  Write-Host "`tContinuing..." -ForegroundColor Green
}


# Check to make sure script is run as administrator
Write-Host "[+] Checking if script is running as administrator.."
$currentPrincipal = New-Object Security.Principal.WindowsPrincipal( [Security.Principal.WindowsIdentity]::GetCurrent() )
if (-Not $currentPrincipal.IsInRole([Security.Principal.WindowsBuiltInRole]::Administrator)) {
  Write-Host "`t[ERR] Please run this script as administrator`n" -ForegroundColor Red
  Write-Host "`n`t`tthis is not the way`n" -ForegroundColor Red
  Read-Host  "Press any key to continue"
  exit
} else {
  Start-Sleep -Milliseconds 500
  Write-Host "`tthis is the way" -ForegroundColor Cyan
  Start-Sleep -Milliseconds 500
}

# Check to make sure Defender is disabled
Write-Host "[+] Checking if Windows Defender service is running.."
$defender = Get-Service -Name WinDefend
if ($defender.Status -eq "Running"){
  Write-Host "`t[ERR] Please disable Windows Defender through Group Policy" -ForegroundColor Red
  Write-Host "`t[+] Hint: https://stackoverflow.com/questions/62174426/how-to-permanently-disable-windows-defender-real-time-protection-with-gpo" -ForegroundColor Yellow
  Write-Host "[-] Do you need to change this setting? Y/N " -ForegroundColor Yellow -NoNewline
  $response = Read-Host
  if ($response -eq "Y") {
    Write-Host "[*] Exiting..." -ForegroundColor Red
    exit
  }else {
    Start-Sleep -Milliseconds 500
    Write-Host "`tLooks good" -ForegroundColor Cyan
    Start-Sleep -Milliseconds 500
  }
}

if ($nochecks -eq $false) {
  
  # Check to make sure Tamper Protection is off
  # This setting is not able to be changed via command line or via scripts
  Write-Host "[+] Checking to make sure Windows Defender Tamper Protection is disabled"
  if (Get-ItemProperty -Path "HKLM:\SOFTWARE\Microsoft\Windows Defender\Features" -Name "TamperProtection") {
    if ($(Get-ItemProperty -Path "HKLM:\SOFTWARE\Microsoft\Windows Defender\Features" -Name "TamperProtection").TamperProtection -ne 0){
    Write-Host "[!] Please disable Windows Defender Tamper Protection and retry install." -ForegroundColor Red
    Write-Host "`t[+] Hint: https://www.tenforums.com/tutorials/123792-turn-off-tamper-protection-windows-defender-antivirus.html" -ForegroundColor Yellow
    Write-Host "[-] Do you need to change this setting? Y/N " -ForegroundColor Yellow -NoNewline
    $response = Read-Host
    if ($response -eq "Y") {
      Write-Host "[*] Exiting..." -ForegroundColor Red
      exit
    }
      Write-Host "`tContinuing..." -ForegroundColor Green
    }
  } else {
    Write-Host "`tTamper Protection is off, looks good." -ForegroundColor Green
  }
  
  # Check to make sure host is supported
  Write-Host "[+] Checking to make sure Operating System is compatible"
  if ((Get-WmiObject -class Win32_OperatingSystem).Version -eq "6.1.7601"){
    Write-Host "Windows 7 is no longer supported. Do you want to continue install? Y/N" -ForegroundColor Yellow
    $response = Read-Host
    if ($response -ne "Y"){
      exit
    }
  }
  
  ## Windows 10 Versions/Build Numbers
  # https://github.com/Disassembler0/Win10-Initial-Setup-Script
  # 21H1                        19043
  # 20H2                        19042
  # 2004 (TBD)                  19041
  # 1909 (November 2019 Update)	18363
  # 1903 (May 2019 Update)    	18362
  # 1809 (October 2018 Update)	17763
  # 1803 (April 2018 Update)   	17134


  $osversion = (Get-WmiObject -class Win32_OperatingSystem).BuildNumber
  $valid_versions = @(18363, 18361, 17763, 17134, 19041, 19042, 19043)
  if ($osversion -notin $valid_versions) {
    Write-Host "`t[ERR] Windows version $osversion is not has not been tested, please use Windows 10 version 1803, 1809, 1903, 1909, 2004, 20H2, or 21H1." -ForegroundColor Yellow
    Write-Host "[-] Do you still wish to proceed? Y/N " -ForegroundColor Yellow -NoNewline
    $response = Read-Host 
    if ($response -ne "Y"){
      exit
    }
  } else {
    Write-Host "`tWindows build $osversion supported." -ForegroundColor Green
  }

  # Check to make sure host has been updated
  Write-Host "[+] Checking if host has been configured with updates"
  if (-Not (get-hotfix | where { (Get-Date($_.InstalledOn)) -gt (get-date).adddays(-30) })) {
    try 
    {
      Write-Host "`t[ERR] This machine has not been updated in the last 30 days, do you want to try installing updates automatically? Y/N " -ForegroundColor Yellow -NoNewline
      $response = Read-Host
      if ($response -eq "Y"){
        Install-PackageProvider -Name NuGet -MinimumVersion 2.8.5.201 -Force
        Install-Module PSWindowsUpdate -Force
        Import-Module PSWindowsUpdate
        Get-WindowsUpdate
        Install-WindowsUpdate -AcceptAll -IgnoreReboot -IgnoreRebootRequired
      } else {
        Write-Host "Please install updates manually." -ForegroundColor Red
        exit
      }
    }
    catch 
    {
      Write-Host "`t[ERR] Could not update automatically, please run Windows Updates manually to continue`n" -ForegroundColor Red
      Read-Host  "Press any key to exit"
      exit
      
    }
    
  } else {
	  Write-Host "`tupdates appear to be in order" -ForegroundColor Green
  }

  #Check to make sure host has enough disk space
  Write-Host "[+] Checking if host has enough disk space"
  $disk = Get-PSDrive C
  Start-Sleep -Seconds 1
  if (-Not (($disk.used + $disk.free)/1GB -gt 58.8)){
    Write-Host "`t[ERR] This install requires a minimum 60 GB hard drive, please increase hard drive space to continue`n" -ForegroundColor Red
    Read-Host "Press any key to continue"
    exit
  } else {
    Write-Host "`t> 60 GB hard drive. looks good" -ForegroundColor Green
  }

  # Prompt user to remind them to take a snapshot
  Write-Host "[-] Do you need to take a snapshot before continuing? Y/N " -ForegroundColor Yellow -NoNewline
  $response = Read-Host
  if ($response -ne "N") {
    Write-Host "[+] Exiting..." -ForegroundColor Red
    exit
  }
  Write-Host "`tContinuing..." -ForegroundColor Green
}

# Get user credentials for autologin during reboots
Write-Host "[+] Getting user credentials ..."
Set-ItemProperty "HKLM:\SOFTWARE\Microsoft\PowerShell\1\ShellIds" -Name "ConsolePrompting" -Value $True
if ([string]::IsNullOrEmpty($password)) {
	$cred=Get-Credential $env:username
} else {
	$spasswd=ConvertTo-SecureString -String $password -AsPlainText -Force
	$cred=New-Object -TypeName "System.Management.Automation.PSCredential" -ArgumentList $env:username, $spasswd
}

if ($nochecks -eq $false){
  Write-Host @"

[!] INSTALL NOTES - PLEASE READ CAREFULLY [!]

- This install is not 100% unattended. Please monitor the install for possible failures. If install
fails you may restart the install by re-running the install script with the following command:

      .\install.ps1 -nochecks 1 [<password>]
                   OR
      .\install.ps1 -nochecks 1 -profile_file .\Profiles\<profile>.json [<password>]

- Please be familiar with these two issues:
https://github.com/fireeye/commando-vm/issues/123
https://github.com/fireeye/commando-vm/issues/122

- Install is not complete until the desktop is cleaned, the readme is placed on the desktop, and the
desktop background is set with the Commando VM logo. For details see 
https://github.com/fireeye/commando-vm/issues/139

- You can check what packages failed install by listing the C:\ProgramData\chocolatey\lib-bad 
directory. Failed packages are stored by folder name. You can attempt manual install with the 
following command:

      cinst -y <package name>


- For any issues please submit to GitHub, find @day1player in the BloodHound Slack, or reach 
out to commandovm@fireeye.com

[!] Please copy these notes for reference [!]
"@ -ForegroundColor Red -BackgroundColor White
  Wait-ForInstall -seconds 60
}

Write-Host "`n[+] Beginning Install...`n" -ForegroundColor Green

Write-Host "[+] Installing Boxstarter"
$rc = installBoxStarter
if ( -Not $rc ) {
	Write-Host "[ERR] Failed to install BoxStarter"
	Read-Host  "      Press ANY key to continue..."
	exit 1
}

# Boxstarter options
$Boxstarter.RebootOk = $true    # Allow reboots?
$Boxstarter.NoPassword = $false # Is this a machine with no login password?
$Boxstarter.AutoLogin = $true   # Save my password securely and auto-login after a reboot
Set-BoxstarterConfig -NugetSources "https://www.myget.org/F/fireeye/api/v2;https://chocolatey.org/api/v2"

# Needed for many applications
# Set up the correct feed
$fireeyeFeed = "https://www.myget.org/F/fireeye/api/v2"
iex "choco sources add -n=fireeye -s $fireeyeFeed --priority 1"
iex "refreshenv"
iex "choco install -y common.fireeye"
iex "refreshenv"

if ($profile -eq $null) {
  # Default install
  Write-Host "[+] Performing normal installation..."
  if (-Not (Set-EnvironmentVariableWrap "FEVM_PROFILE" "Commando_Default")) {
    Write-Warning "Failed to set environment variable FEVM_PROFILE"
  }
  choco upgrade -y common.fireeye
  if ([System.Environment]::OSVersion.Version.Major -eq 6) {
    Install-BoxstarterPackage -PackageName commandovm.win7.installer.fireeye -Credential $cred
    Install-BoxStarterPackage -PackageName commandovm.win7.config.fireeye  -Credential $cred
  } elseif ([System.Environment]::OSVersion.Version.Major -eq 10) {    
    choco config set cacheLocation ${Env:TEMP}
    iex "choco upgrade -y commandovm.win10.preconfig.fireeye"
    Install-BoxstarterPackage -PackageName commandovm.win10.installer.fireeye -Credential $cred
    Install-BoxStarterPackage -PackageName commandovm.win10.config.fireeye  -Credential $cred
  }
  exit 0
} 

# The necessary basic environment variables
$EnvVars = @(
  "VM_COMMON_DIR",
  "TOOL_LIST_DIR",
  "TOOL_LIST_SHORTCUT",
  "RAW_TOOLS_DIR",
  "FEVM_PROFILE"
  )

foreach ($envVar in $EnvVars) {
  try {
		$value = [Environment]::ExpandEnvironmentVariables($profile.env.($envVar))
		if (-Not (Set-EnvironmentVariableWrap $envVar $value)) {
			Write-Warning "[ - ] Failed to set environment variable $envVar"
		}
  } catch {}
}

$PackageName = "MyInstaller"
$TemplateDir = $profile.env.TEMPLATE_DIR
$Packages = $profile.packages
Make-InstallerPackage $PackageName $TemplateDir $Packages
Invoke-BoxStarterBuild $PackageName
if ([System.Environment]::OSVersion.Version.Major -eq 6) {
  Install-BoxStarterPackage -PackageName $PackageName -Credential $cred
  Install-BoxStarterPackage -PackageName commandovm.win7.config.fireeye  -Credential $cred
} elseif ([System.Environment]::OSVersion.Version.Major -eq 10) {
  iex "choco upgrade -y commandovm.win10.preconfig.fireeye"
  Install-BoxStarterPackage -PackageName $PackageName -Credential $cred
  Install-BoxStarterPackage -PackageName commandovm.win10.config.fireeye  -Credential $cred
}
exit 0
